parcels/particle.py

```diff
@@ -15,25 +15,55 @@
 def AdvectionRK4(particle, grid, time, dt):
     f_lat = dt / 1000. / 1.852 / 60.
     f_lon = f_lat / math.cos(particle.lat*math.pi/180)
-    u1 = grid.U[time, particle.lon, particle.lat]
-    v1 = grid.V[time, particle.lon, particle.lat]
-    lon1, lat1 = (particle.lon + u1*.5*f_lon, particle.lat + v1*.5*f_lat)
-    u2, v2 = (grid.U[time + .5 * dt, lon1, lat1], grid.V[time + .5 * dt, lon1, lat1])
-    lon2, lat2 = (particle.lon + u2*.5*f_lon, particle.lat + v2*.5*f_lat)
-    u3, v3 = (grid.U[time + .5 * dt, lon2, lat2], grid.V[time + .5 * dt, lon2, lat2])
-    lon3, lat3 = (particle.lon + u3*f_lon, particle.lat + v3*f_lat)
-    u4, v4 = (grid.U[time + dt, lon3, lat3], grid.V[time + dt, lon3, lat3])
+    posvertdir = -1.  # TODO: Decide on direction of positive w
+    u1 = grid.U[time, particle.lon, particle.lat, particle.dep]
+    lon1 = particle.lon + u1*.5*f_lon
+    v1 = grid.V[time, particle.lon, particle.lat, particle.dep]
+    lat1 = particle.lat + v1*.5*f_lat
+    if hasattr(grid, 'W'):
```

Comment:
I think for the moment, separate RK4_2D and RK4_3D methods is the way to
go. The user will have to choose the appropriate kernel, and we can put a
single check before the main loop of `ParticleSet.execute` to ensure that
the grid has the appropriate dimensions.

On 1 May 2016 at 14:55, Michael Lange <notifications@github.com> wrote:

> In parcels/particle.py
> <https://github.com/OceanPARCELS/parcels/pull/64#discussion_r61679522>:
>
> > @@ -15,25 +15,55 @@
> >  def AdvectionRK4(particle, grid, time, dt):
> >      f_lat = dt / 1000. / 1.852 / 60.
> >      f_lon = f_lat / math.cos(particle.lat*math.pi/180)
> > -    u1 = grid.U[time, particle.lon, particle.lat]
> > -    v1 = grid.V[time, particle.lon, particle.lat]
> > -    lon1, lat1 = (particle.lon + u1*.5*f_lon, particle.lat + v1*.5*f_lat)
> > -    u2, v2 = (grid.U[time + .5 * dt, lon1, lat1], grid.V[time + .5 * dt, lon1, lat1])
> > -    lon2, lat2 = (particle.lon + u2*.5*f_lon, particle.lat + v2*.5*f_lat)
> > -    u3, v3 = (grid.U[time + .5 * dt, lon2, lat2], grid.V[time + .5 * dt, lon2, lat2])
> > -    lon3, lat3 = (particle.lon + u3*f_lon, particle.lat + v3*f_lat)
> > -    u4, v4 = (grid.U[time + dt, lon3, lat3], grid.V[time + dt, lon3, lat3])
> > +    posvertdir = -1.  # TODO: Decide on direction of positive w
> > +    u1 = grid.U[time, particle.lon, particle.lat, particle.dep]
> > +    lon1 = particle.lon + u1*.5*f_lon
> > +    v1 = grid.V[time, particle.lon, particle.lat, particle.dep]
> > +    lat1 = particle.lat + v1*.5*f_lat
> > +    if hasattr(grid, 'W'):
>
> Ok, as far as I can see we have 3 potential solutions here:
> 1. Bite the bullet and implement vector fields / vector variables. This
> would mean we sample the field with something like vel =
> grid.Velocity[time, particle.position], where grid.Velocity is {U, V} in
> 2D and {U, V, W} in 3D. Note that vel and particle.position are now
> vector variables with 2 or 3 components according to grid dimension.
> 2. Simply replicate the kernels and write out AdvectionRK4_2D and
> AdvectionRK4_3D. Note, that might require us to do some sanity checking
> about whether a kernel is applicable, given a particular grid. This, in
> turn, will require us to store some meta-information with kernels, such
> "local variables used" or "fields sampled", which we can get out of the JIT
> analysis engine.
> 3. Assume every grid has W and create a kind of "dummy field" that always
> returns zeros for 2D runs. As mentioned before this will be wasteful in
> terms of performance, and will prevent us running any meaningful benchmarks
> for a while.
>
> Now, 1 is clearly "the right thing to do" in the long run, since we are
> trying abstract grid assumptions, such as the U-V-W split of velocity, away
> from the user. However, it is also by far the most complicated to implement
> and might take a significant portion of someone's time, so I'm ok with
> putting in one of the other quick fixes first, as long as we are all aware
> this needs to happen eventually. I'm also open to alternatives, of course...
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/OceanPARCELS/parcels/pull/64/files/4a5eebc2d05db4f499d263402568d47ab4a82ebb#r61679522>
>